### ISSUE TITLE
Newlines as whitespace text

### ISSUE THREAD

**Issue by RReverser** [facebook/jsx#19]
Currently contents like

``` html
<b>{this.title}</b> <span>123</span>
```

are converted to

``` javascript
React.DOM.b(null, this.title), " ", React.DOM.span(null, "123")
```

which is logical and expected since whitespace might change the layout (i.e. inside element with `white-space: pre` and in other cases).

However:

``` html
<pre>
    <b>{this.title}</b> <span>123</span>
</pre>
```

is converted to

``` javascript
React.DOM.pre(null, 
    React.DOM.b(null, this.title), " ", React.DOM.span(null, "123")
)
```

so whitespaces across newlines are inserted only in code, but not treated as regular text, which means we loose them completely in generated layout.

Shouldn't we treat them as text too?

**Comment by garthk** [comment]
I'd like to see whitespace inserted on all newlines that aren't between tag boundaries. It's least surprising, and (unless I'm missing something) can be done without harm to the list of edge cases above.

Consider:

```
<p>
a
b
c
<p>
```

yields:

```
React.createElement("p", null, 
"a" + ' ' +
"b" + ' ' +
"c"
)
```

Given that, I'm shocked when the spaces disappear when I replace one line in the text with a tag:

```
<p>
a
<a>b</a>
c
</p>
```

yields:

```
React.createElement("p", null, 
"a", 
React.createElement("a", null, "b"), 
"c"
)
```

What harm am I missing, apart from possibly lending moral support to an HTML feature @syranide calls "[horrible](https://github.com/facebook/react/pull/480#issue-22160691)" in #480, that would come if we preserved whitespace for the following newlines?

* between text and begin-tags, e.g. `text\n<div>`
* between end-tags and text, e.g. `</div>\ntext`

**Comment by syranide** [comment]
@garthk I'm not a fan of whitespace rules for text at all, to lend some more context #28 (and #8 #35).

> ... would come if we preserved whitespace for the following newlines?

If you were to have that there would no longer be a way *not* to have whitespace between text and tags (except for making the entire string an expression), whereas the current way allows you to easily add whitespace you think it should be there and you'll also be explicit for others benefit. The verbosity and side-effects of `{' '}`is undesirable though (hence why I'd like to see #28 or whatever) but the alternative is far worse in my opinion.

Inserting whitespace between text lines makes quite a lot of practical sense, it's almost certainly what you wanted as you break on space when lines get too long and if you don't like it you can just put `{}` at the end and it's gone. Whereas `foo\n<tag>bar</tag>` having a whitespace but not `<b>foo</b>\n<tag>bar</tag>` feels ambigious and means that using newlines to structure your code affects the outcome of it (which is what I think is *horrible*).

I think it's important to remember, especially now with React-Native, that HTML whitespace rules makes *absolutely no sense* for frontends other than HTML, JSX is not a markup language.

tl;dr I don't like the whitespace rules, but IMHO it's the best compromise. I think an alternative (not strictly HTML) approach is necessary.

**Comment by facebook-github-bot-7** [comment]
Thank you for reporting this issue and appreciate your patience. We've  notified the core team for an update on this issue. We're looking for a response within the next 30 days or the issue may be closed.

**Comment by LucaColonnello** [comment]
Hello guys,
almost 3 years from the last comment on this issue, we have the same problem and we are still wondering what to do on the transformation step.

The problem though is real https://github.com/babel/babel/issues/7360 and affects other parsers as well https://github.com/Microsoft/TypeScript/issues/22186.

We really need to have a proper solution for this. It doesn't have to address all the issues above (I appreciate the downsides of a solution over another) but we need a direction so that at least all the parsers are following the spec.
As @sebmarkbage said, this is unspecified, but it would be really helpful to have a spec for it.

A suggestion I would give, is to implement the same behavior as the HTML standard does.


### TL;DR

**Don't strip out whitespaces and tab if they are following text nodes.**


### Text before newline and tabs

<img width="1676" alt="screen shot 2018-02-28 at 10 50 46" src="https://user-images.githubusercontent.com/536973/36783871-608c8224-1c75-11e8-92c0-8467471086e8.png">

**There's some text before the newline, so the browser carry over the newline and tabs.**


### Tag before newline and tabs

<img width="1680" alt="screen shot 2018-02-28 at 10 51 25" src="https://user-images.githubusercontent.com/536973/36783916-90049cc6-1c75-11e8-9727-8757b20f184a.png">

**There's no text before the newline, only the strong tag, so the browser treat it as indentation and strips eventual newlines, whitespaces and tabs out.**

---

In both cases the browser renders the whitespace, which means that we only need to transform to what is standard on the HTML spec in order to reach the expected behavior on the renderer layer.

I'm aware that JSX is not only for HTML, but I found sensible that if there's a text before the newline, the following characters (newlines, whitespaces and tabs) are considered as part of the above text.

**Comment by LucaColonnello** [comment]
Also note the HTML spec mentioning the same here https://www.w3.org/TR/html4/appendix/notes.html#notes-line-breaks

**Comment by Strate** [comment]
Just ran into this issue trying to use native bootstrap 4 and react :(

**Comment by cancerberoSgx** [comment]
@syranide Thanks! the trick with {' '} did work! I however don't agree. Most of you are assuming that everybody targets HTML DOM or a visual medium with stylesheets. 

In my case I'm targeting the command line and there we don't have much more than spaces for positioning. 

Even with some layouts implementations, I want to implement a text area like or widget without my users to be forced to manually write those {' '} . Is not a problem for the en user but it is for document authors. 

BTW HTML makes exceptions for these cases in text area and pre, which seems to be reasonable / useful. Any idea idea of how to support that ? My only idea now is to preprocess the JSX to add the {' '} since I don't have any control on that part of the code generation. Or there is some way of hooking in there ?

Thank you very much.

**Comment by syranide** [comment]
@cancerberoSgx JSX inline strings IMHO are only really well-suited to small snippets of text/labels. If you want something more specific or extensive then I recommend putting it all into a string, and maybe not even as JSX-children but as a prop. I.e. `<FooBar>{"   fooo " + index + "baaar  "}</FooBar>` or `<FooBar text={"   fooo " + index + "baaar  "} />`, or you can use ES6 template-strings if you want nicer interpolation or multi-line strings (and don't mind the indentation).

**Comment by cancerberoSgx** [comment]
@syranide Aha, thanks for your view, My relationship with jsx lately was trying to implement renderers targeting different mediums (like a string, latex, CLI, etc and in general those mediums are just buffers / strings. 

As a language, that has a duality with JavaScript / TypeScript I think is important not to limit its design to accommodate to *most used* use cases. For me is not just another template language, is an Object representation alternative syntax! specially useful for visual ouptuts. 

And in that matter, I disagree with you: **JSX.TextNode** is the most practical and unique kind of Node that JSX introduces. It support cases that cannot be supported in that natural way.  (even string templates where you need to wrap and escape but worst , you cannot represent structure in a string literal). 

Imagine the user is authoring a book, or article or news paper and is forced to use strings or attribtues for that !!!. JSX.TextNode is the natural way of accomodate such content and no other kind of node in JS/TS/JSON is better for that. The other kind of nodes, like JSXElement, TagName, JSXAttributes, for me, have more straight forward equivalence with JavaScript node kinds, but JSXTextNode, from the point of view of authors has unique capability syntax.

Just wanted to give this opinion since I'm surprised how lightly the argument is to justify (defacto) messing with  JSX.TextNode and not providing any means of configuration. JSX.TextNode also is important for parent-children semantics declarations. WOULD BE sad that the language ends up being just elements with attributes because of HTML developers peacefulness (and I one of them). Then I prefer to represent that pattern with just JS objects... those were designed for that and not for text.

**Comment by syranide** [comment]
> Imagine the user is authoring a book, or article or news paper and is forced to use strings or attribtues for that !!!. JSX.TextNode is the natural way of accomodate such content and no other kind of node in JS/TS/JSON is better for that.

@cancerberoSgx It could also be a matter of `<Foobar> Hello </Foobar>` vs something like `<Foobar>{' Hello '}</Foobar>`. As the second relies on JS-strings with boundary markers it's very precise (or rather, it's just an expression), and you would interpolate strings like you normally do.

But the first thing to realize I would say is that JSX is not a syntax for writing blogs or articles, it's for structuring data, static button labels, etc. It's just an extension to make it less cumbersome and more readable than what you can otherwise achieve with the existing JS syntax. Content data should be stored in a database or such and loaded into React at run-time. Also, as you develop you generally tend to want to support different languages, so then even button labels become function calls and you pretty much stop using JSX inline strings entirely.

IMHO, JSX is not HTML, JSX is an extension of JS, which is about code, not arbitrary text representation. It's about providing a more readable syntax for specifying a type + data. If you want to encode markdown in JS today you need to put it in an ES6 template-string and process it at run-time, or use some markdown-file transform, this is no different.

> you cannot represent structure in a string literal).

ES6 template-strings support that though and will encode all white-space contained within it, for better or worse. It's the right tool for the job if you want more exact run-time control.

> Just wanted to give this opinion since I'm surprised how lightly the argument is to justify (defacto) messing with JSX.TextNode and not providing any means of configuration.

It's not lightly, it's actually me who designed the white-space rules way back. The fact is that when you evaluate the possibilities it essentially boils down to two sensible paths.

1. Either you keep all white-space, indentation, newlines, etc. From beginning to end. However, this is problematic, because not only is indentation then encoded into the generated code. Trimming white-space then becomes a run-time concern (which would actually be quite complicated and expensive to get right) and you will end up with tons and tons of white-space-only strings everywhere in the final code, and worse run-time behavior would be affected by code structure (i.e. indentation and newlines). Also, this would have a universally detrimental effect on not only code-size, but also performance.

2. Trim all white-space and treat inline strings more like tokens. Then you get a minimal representation which is well-suited to basic text representation, i.e. labels and such. I've advocated for removing the HTML-entity parsing rules and using JS escape sequences instead, but there's been little interest or movement on that it seems. That's one way to "solve" the white-space issues by way of just writing `\ ` or such for forcing leading/trailing white-space and you could then explicitly insert newlines with e.g. `\n`. Not perfect, but all use-cases would be reasonably covered, but there doesn't exist a perfect catch-all solution, there's always a trade-off.

As JSX is a compile-time transform there is really no sensible means of configuration without resorting to directives, but directives are messy. But if you want to have a babel-jsx plugin with different white-space rules then by all means.

**Comment by vjeux** [comment]
If you’re writing a book, MDX might be more suited than straight JSX. https://mdxjs.com

**Comment by cancerberoSgx** [comment]
@syranide thanks

**Comment by nik72619c** [comment]
Hey folks,
Saw one wierd catch in the jsx, but not sure if its relevant to this or not
**Note:- `<spacebar>` means a whitespace applied using spacebar key in the keyboard**

```jsx
<span>
  <span>this is text1<spacebar></span>
  <span>and text2</span>
</span>
```

renders 👉🏻 `this is text1 and text2`

But,
```jsx
<span>
  <span>
    this is text1<spacebar>
  </span>
  <span>
    and text2
  </span>
</span>
```

renders  👉🏻 `this is text1and text2`

Could someone please guide me on this?
@syranide @RReverser @sebmarkbage 🙂 
Thanks!

**Comment by syranide** [comment]
@nik72619c That is correct, all lines are trimmed. Since there are no wrapping elements in the second example the whitespace at the end gets trimmed. Here's the layman's explanations of the algorithm: https://github.com/facebook/react/pull/480#issuecomment-31296039

You can force a whitespace with `{' '}`, `{'this is text1 '}` or similar.

**Comment by nik72619c** [comment]
ah got it 👍 
Thanks again!

**Comment by syranide** [comment]
@nik72619c Anything really, simply put, like the explanation says, leading and trailing white-space is trimmed (from the point-of-view of the source code lines). So white-space that is surrounded by **anything** (text, tag, braces) will be kept, otherwise it will be removed.
